### PR TITLE
[SYCL][CUDA] Testing for use of CUDA primary context

### PR DESCRIPTION
**Pull request by steffenlarsen**
Signed-off-by: Steffen Larsen <steffen.larsen@codeplay.com>

### CODE REVIEW

File: sycl/test/backend/cuda/primary_context.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+int main() {
+  try {
+    context c;
+  } catch (device_error e) {
+    std::cout << "Failed to create device for context" << std::endl;
+  }
+
+  device deviceA = default_selector().select_device();
```
Comment (romanovvlad) [inline sycl/test/backend/cuda/primary_context.cpp]: Please don't use default selector, it's free to return any other device. I suggest writing device selector which selects nvidia device.

File: sycl/test/backend/cuda/primary_context.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    assert(cudaContext == primaryCudaContext);
+
+    cuDevicePrimaryCtxRelease(cudaDevice);
+  }
+  {
+    std::cout << "create multiple contexts for one device" << std::endl;
+    context Context1(deviceA, async_handler{}, true);
+    context Context2(deviceA, async_handler{}, true);
+
+    CUcontext cudaContext1 =
```
Comment (romanovvlad) [inline sycl/test/backend/cuda/primary_context.cpp]: Could you please use the same naming style for all variables: Context2 vs cudaContext1

File: sycl/test/backend/cuda/primary_context.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
+// REQUIRES: cuda
+// RUN: %clangxx %s -I%opencl_include_dir -I%cuda_toolkit_include -o %t.out -lcuda -lsycl
```
Comment (AlexeySachkov) [inline sycl/test/backend/cuda/primary_context.cpp]: Why do you need `-I%opencl_include_dir` in this test?

Comment (steffenlarsen) [reply sycl/test/backend/cuda/primary_context.cpp]: Without it, the OpenCL header supplied with the CUDA toolkit is picked up which causes problems as it is behind OpenCL 2.1

Comment (AlexeySachkov) [reply sycl/test/backend/cuda/primary_context.cpp]: Oh, it seems to be needed for the SYCL runtime, not for the test

I think it worth to put this comment into directly into the test - someone not familiar with the issue might step into it while adding new tests

File: sycl/test/backend/cuda/primary_context.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+void check(bool condition, const char *conditionString, const char *filename,
+           const long line) noexcept {
+  if (!condition) {
+    std::cerr << "CHECK failed in " << filename << "#" << line << " "
+              << conditionString << "\n";
+    std::abort();
+  }
+}
+
+#define CHECK(CONDITION) check(CONDITION, #CONDITION, __FILE__, __LINE__)
```
Comment (steffenlarsen) [inline sycl/test/backend/cuda/primary_context.cpp]: `CHECK` borrowed from https://github.com/intel/llvm/pull/1200. Once it is merged I will use that implementation instead.

File: sycl/test/backend/cuda/primary_context.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "create single context" << std::endl;
+    context Context(deviceA, async_handler{}, true);
+
+    CUdevice cudaDevice = reinterpret_cast<pi_device>(deviceA.get())->get();
+    CUcontext cudaContext = reinterpret_cast<pi_context>(Context.get())->get();
+
+    CUcontext primaryCudaContext;
+    cuDevicePrimaryCtxRetain(&primaryCudaContext, cudaDevice);
+
+    assert(cudaContext == primaryCudaContext);
```
[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/test/backend/cuda/primary_context.cpp]: Good call. It relates to another open PR in fact. See https://github.com/intel/llvm/pull/1174/files#r386428677.

File: sycl/test/backend/cuda/primary_context.cpp (hunk 1 above)
Comment (steffenlarsen) [reply sycl/test/backend/cuda/primary_context.cpp]: With SYCL_BE=PI_CUDA it should be alright, but you are right that it is probably safer in the long run to use a custom selector.

File: sycl/test/backend/cuda/primary_context.cpp (hunk 3 above)
Comment (steffenlarsen) [reply sycl/test/backend/cuda/primary_context.cpp]: Good idea. I've added a note about this.

**Comment by bader** [comment]
@steffenlarsen, please, resolve merge conflicts.

**Comment by steffenlarsen** [comment]
> @steffenlarsen, please, resolve merge conflicts.

Thank you for the reminder. The conflicts have been fixed.

**Pull request merged**